Warns

hunse opened the pull request:
The `recwarn` function argument is useful, but sometimes I just want to make sure that a warning of a particular type is thrown, and this takes a number of lines with `recwarn`. I added a `warns` fixture to ensure that warnings are thrown. Similar to `raises`, but for warnings.

nicoddemus commented on CHANGELOG:
```diff
@@ -120,6 +120,9 @@
 2.7.3 (compared to 2.7.2)
 -----------------------------
 
+- Add 'warns' to assert that warnings are thrown (like 'raises').
```
I think this should be in the changelog for `2.8`.

hunse replied:
Oh yeah, sorry, I rebased and forgot to move it.

nicoddemus commented on _pytest/python.py:
```diff
@@ -1163,6 +1165,56 @@ def __exit__(self, *tp):
+    def __exit__(self, type, value, traceback):
+        self.catcher.__exit__(type, value, traceback)
+        if not any(r.category in self.ExpectedWarning for r in self.record):
+            pytest.fail("DID NOT RAISE")
```
`"DID NOT WARN"` perhaps? :wink:

hunse replied:
Haha, yeah, good idea :smile:

nicoddemus commented on _pytest/python.py:
```diff
@@ -1163,6 +1165,56 @@ def __exit__(self, *tp):
 
+# builtin pytest.warns helper
+
+def warns(ExpectedWarning, *args, **kwargs):
```
Minor, but perhaps this should be `expected_warning`?

hunse replied:
I was just following the style of `raises`, above. This function definitely needs a docstring, though.

nicoddemus replied:
Oh OK, I didn't notice.

nicoddemus commented on testing/python/warns.py:
```diff
@@ -0,0 +1,43 @@
+                          (RuntimeWarning, SyntaxWarning),
+                          lambda: warnings.warn('w3', UserWarning)))
+
+    @pytest.mark.skipif('sys.version < "2.5"')
```
This is not necessary as pytest supports `2.6+` only. :smile:

hunse replied:
Oh, sweet! Again, just copying from `raises.py`.

nicoddemus commented:
Thanks for the PR @hunse! :smile:

nicoddemus commented:
Hmmm could you write some docs for `pytest.warns`? I think you might add it to `assert.txt`... make sure to add a `.. versionadded:: 2.8` directive as well. :smile:

hunse commented:
Okay, doc is added.

nicoddemus commented on doc/en/assert.txt:
```diff
@@ -113,6 +113,38 @@ exceptions your own code is deliberately raising, whereas using
+        warnings.warn("another warning", RuntimeWarning)
+
+    assert len(record) == 1  # check that only one warning was raised
+    assert record[0].message.message == "another warning"
```
`message.message`? Is this a typo?

hunse replied:
No, that's actually the way you have to do it, because of how `warnings.WarningMessage` is set up. `WarningMessage.message` returns the actual exception, not the message, so you have to call `message` again. Though your question made me realize that in Python3, warnings no longer have a `message` attribute, just an `args` one, so I've changed things to work in Python3 as well.

nicoddemus commented on doc/en/assert.txt:
```diff
@@ -113,6 +113,38 @@ exceptions your own code is deliberately raising, whereas using
+    assert len(record) == 1  # check that only one warning was raised
+    assert record[0].message.message == "another warning"
+
+Alternatively, you can examine raised warnings in detail using the ``recwarns``
```
I think this paragraph should be moved to after the next one (`"You can also call..."`), plus it would be nice to contain a link to the `recwarns` docs. :smile:

nicoddemus commented:
Nice @hunse, much appreciated! 

If nobody has any more comments, I will merge this tomorrow. :smile:

hunse commented:
Okay, changed the docs around and added the link to `recwarn`. The only thing I'm worried about now is that the `recwarn` [page](https://pytest.org/latest/recwarn.html) says it's about "Asserting deprecation and other warnings", which makes it sound like it's the way to do that, and it's always confusing when there are two ways to do things. Should the `recwarn` page be rewritten so it's more about recording warnings? I'm not quite sure how I would word it, though.

nicoddemus commented:
Hmm good point. I see two options:

* Add a link back to `pytest.warns` in `recwarn`, explaining its intended use for simple cases;
* Move `pytest.warns` docs to `recwarn` and keep it all in the same place;

I after your points I'm thinking the latter. What do you think?

hunse commented:
Yeah, I'm thinking the same thing. I was also wondering if it would be good to have the same interface for warnings recorded by `recwarn` and those recorded by `pytest.warns`.

hunse commented:
Though it might be nice to have something after `pytest.raises` that at least points to the `recwarn` document. I think a lot of people might read through `pytest.raises` and then think "but how do I do this for warnings?"

nicoddemus commented:
>  I was also wondering if it would be good to have the same interface for warnings recorded by recwarn and those recorded by pytest.warns.

I think so, but don't know how difficult that would be to implement.

> I think a lot of people might read through pytest.raises and then think "but how do I do this for warnings?"

Agreed. What about a one liner after `pytest.raises` like: `"To capture warnings, take a look at <link to recwarn>"`?

hunse commented:
Okay, `warns` and `recwarn` use the same code, now. It seems a lot cleaner. @nicoddemus, let me know what you think, and I'll update the documentation.

nicoddemus commented on _pytest/recwarn.py:
```diff
@@ -13,83 +16,177 @@ def pytest_funcarg__recwarn(request):
-        warnings.warn_explicit = oldwarn_explicit
-        warnings.warn = oldwarn
-    if not l:
-        __tracebackhide__ = True
```
Keep `__tracebackhide__` in the body of the function, this way when it fails the traceback will point to the point of the call, not to the internal details of this function.

nicoddemus commented on _pytest/recwarn.py:
```diff
@@ -13,83 +16,177 @@ def pytest_funcarg__recwarn(request):
+            wcheck.finalize()
+    else:
+        func = args[0]
+        wcheck.start()
```
Minor, but you can replace this with:

```python
with wcheck:
    return func(*args[1:], **kwargs)
```

Same for the lines right above this one.

hunse replied:
Good point. I was still in the old mindset of trying to have this work without context managers, since that was in `pytest.raises`, but you're right that that's no longer necessary since we only support 2.6+.

nicoddemus commented on _pytest/recwarn.py:
```diff
@@ -1,7 +1,10 @@
+import pytest
 
 
 def pytest_funcarg__recwarn(request):
```
Since we're at it, could you please replace this with:

```python
@pytest.fixture
def recwarn(request):
```

:smile: ?

nicoddemus commented:
@hunse excellent work, thanks for the nice refactoring! :smile:

nicoddemus commented:
About the docs, I think it makes sense to add `pytest.warns` together with other documentation related to warnings in `recwarn.txt`, what do you think?

hunse commented:
Agreed.

nicoddemus commented on _pytest/recwarn.py:
```diff
@@ -13,83 +17,171 @@ def pytest_funcarg__recwarn(request):
-        def reset_filters():
-            warnings.filters[:] = oldfilters
-        request.addfinalizer(reset_filters)
     wrec = WarningsRecorder()
```
Minor: this fits nicely as a yield_fixture:

```python
@pytest.yield_fixture
def recwarn():
    wrec = WarningsRecorder()
    with wrec:
        warnings.simplefilter('default')
        yield wrec
```

 (should've seen it before, sorry! :sweat_smile:)

hunse replied:
Oh wow, that's cool. I didn't even know that was possible!

hunse commented on _pytest/recwarn.py:
```diff
@@ -13,83 +17,164 @@ def pytest_funcarg__recwarn(request):
+        return iter(self._list)
+
+    def __len__(self):
+        return len(self._list)
```
@nicoddemus: how would I add documentation for these kinds of methods in `recwarn.txt`? There seems like there should be a clearer way than just adding them as `.. method::`.

nicoddemus replied:
You can use `auto` functions from sphinx... you can see them in action [here](https://github.com/pytest-dev/pytest/blob/master/doc/en/writing_plugins.txt#L497).

nicoddemus commented:
@hunse could you fix the flakes error? After this I think we can merge this. :smile:

hunse commented:
I have no idea why it's getting that error. The Travis build says it ran on bd3877b, which is not the most recent update. Can you force it to rerun? There's no reason that it should not be able to import `WarningsChecker` from `raises.py`; in fact, the file that it's complaining about runs fine, i.e. it **can** import `WarningsChecker`. It doesn't make any sense because that file must have run as part of the test suite, which passed, so I don't know why flakes is finding problems.

nicoddemus commented:
Oh I didn't even check the error... it seems like travis misinterpreted some rebase done in the branch. I tried re-submitting the job, but it insists on running against commit bd3877b...

Anyway, I just tested it and locally and everything passes, so we can ignore that error. :smile:

nicoddemus commented:
Thanks a lot for the work @hunse! 

Cheers,
:beers:

The-Compiler commented:
Hm - it seems the same error shows up on Travis now that it's [merged to master](https://travis-ci.org/pytest-dev/pytest/jobs/73990934).

nicoddemus commented:
Oh dear... I will take a look at this later then. No idea why flakes is acting up (if anybody wants to take a shot a it, please be my guest);

hunse commented:
The other thing that is very confusing (and makes it very annoying to debug) is that this error doesn't seem to appear when running `flakes` locally. At least I didn't have it appear on my machine, and @nicoddemus didn't see it either. Maybe this is a bug in `flakes` that was fixed, and Travis just needs an update?

The-Compiler commented:
Travis uses tox for the tests, so it should use a recent pyflakes version.

Maybe try running `git clean -xdf` to clean up any cache files etc.? (Make sure there's nothing important untracked in `git status` first)

The-Compiler commented:
I can actually reproduce this when doing `-i ALL=https://devpi.net/hpk/dev/` - in my experience, devpi sometimes fails to pick up new upstream versions, so I'm guessing that's what's happening...

@hpk42 - can you force an update? Can we somehow make tox only use devpi for the needed packages (pylib I guess?) and use PyPI directly for everything else?

nicoddemus commented:
As a workaround, I moved `import WarningsChecker` to a local import and then `flakes` passed, but we should update flakes on `devpi` when we can.

hpk42 commented:
Seems like this PR broke issue #1037
